### ISSUE TITLE
RuntimeWarning:: invalid value encountered in cast

### ISSUE THREAD

**Issue by noaft** [sberbank-ai/Real-ESRGAN#42]
This is my code:
![image](https://github.com/user-attachments/assets/e271bf69-89e3-4f0a-8e46-2ed603fb460f)
i was "pip install git+https://github.com/sberbank-ai/Real-ESRGAN.git" and don't have problem. But when using have warning RuntimeWarning:  invalid value encountered in cast. It means none valid in predict.

**Comment by smuggypusss** [comment]
Yeah I am also seeing the same issue here

**Comment by smuggypusss** [comment]
I got it, run the model in FP32 instead of FP16, the error was in the patch 0 after the inference, force the model output to a valid range, after some debugging I saw that the patch 0 had Nan values.

this is the code that worked for me, this is model.py


`import os
import torch
from torch.nn import functional as F
from PIL import Image
import numpy as np
import cv2
from huggingface_hub import hf_hub_url, cached_download

import matplotlib.pyplot as plt

from .rrdbnet_arch import RRDBNet
from .utils import pad_reflect, split_image_into_overlapping_patches, stich_together, \
                   unpad_image


HF_MODELS = {
    2: dict(
        repo_id='sberbank-ai/Real-ESRGAN',
        filename='RealESRGAN_x2.pth',
    ),
    4: dict(
        repo_id='sberbank-ai/Real-ESRGAN',
        filename='RealESRGAN_x4.pth',
    ),
    8: dict(
        repo_id='sberbank-ai/Real-ESRGAN',
        filename='RealESRGAN_x8.pth',
    ),
}


class RealESRGAN:
    def __init__(self, device, scale=4):
        self.device = device
        self.scale = scale
        self.model = RRDBNet(
            num_in_ch=3, num_out_ch=3, num_feat=64, 
            num_block=23, num_grow_ch=32, scale=scale
        )
        
    def load_weights(self, model_path, download=True):
        if not os.path.exists(model_path) and download:
            assert self.scale in [2,4,8], 'You can download models only with scales: 2, 4, 8'
            config = HF_MODELS[self.scale]
            cache_dir = os.path.dirname(model_path)
            local_filename = os.path.basename(model_path)
            config_file_url = hf_hub_url(repo_id=config['repo_id'], filename=config['filename'])
            cached_download(config_file_url, cache_dir=cache_dir, force_filename=local_filename)
            print('Weights downloaded to:', os.path.join(cache_dir, local_filename))
        
        loadnet = torch.load(model_path)
        if 'params' in loadnet:
            self.model.load_state_dict(loadnet['params'], strict=False)
        elif 'params_ema' in loadnet:
            self.model.load_state_dict(loadnet['params_ema'], strict=False)
        else:
            self.model.load_state_dict(loadnet, strict=True)
        self.model.eval()
        self.model.to(self.device)

    def predict(self, lr_image, batch_size=4, patches_size=192,
                padding=24, pad_size=15):
        scale = self.scale
        device = self.device
        lr_image = np.array(lr_image)
        lr_image = pad_reflect(lr_image, pad_size)

        patches, p_shape = split_image_into_overlapping_patches(
            lr_image, patch_size=patches_size, padding_size=padding
        )
        print("Checking patches before model inference:")
        for i, patch in enumerate(patches):
            print(f"Patch {i} min: {patch.min()} max: {patch.max()} nan count: {np.isnan(patch).sum()}")
        print(f"Number of patches: {len(patches)}, Patch shape: {patches[0].shape}")

        img = torch.FloatTensor(patches/255).permute((0,3,1,2)).to(device).detach()
        img = torch.clamp(img, 0, 1)
        with torch.no_grad():
            res = self.model(img[0:batch_size])
            res = torch.nan_to_num(res, nan=0.0)  # Convert NaNs to 0
            res = torch.clamp(res, 0, 1)  # Force values to be in range
            print("Patch 0 before model min:", img[0].min().item(), "max:", img[0].max().item())
            print("Patch 0 after model min:", res.min().item(), "max:", res.max().item())
            for i in range(batch_size, img.shape[0], batch_size):
                patch_output = self.model(img[i:i + batch_size])
                print(f"Patch {i} min:", patch_output.min().item(), "max:", patch_output.max().item())  # New debug line
                res = torch.cat((res, patch_output), 0)

        print(f"Processed {res.shape[0]} patches out of {len(patches)} expected")

        res = torch.nan_to_num(res, nan=0.0, posinf=1.0, neginf=0.0)  # Replace NaNs
        sr_image = res.permute((0, 2, 3, 1)).clamp(0, 1).cpu()

        np_sr_image = sr_image.numpy()
        print("NaN count:", np.isnan(np_sr_image).sum())
        print("Inf count:", np.isinf(np_sr_image).sum())

        padded_size_scaled = tuple(np.multiply(p_shape[0:2], scale)) + (3,)
        scaled_image_shape = tuple(np.multiply(lr_image.shape[0:2], scale)) + (3,)
        np_sr_image = stich_together(
            np_sr_image, padded_image_shape=padded_size_scaled, 
            target_shape=scaled_image_shape, padding_size=padding * scale
        )
        print("Stitched image shape:", np_sr_image.shape)
        plt.imshow(np_sr_image)
        plt.show()
        np_sr_image = np.nan_to_num(np_sr_image, nan=0.0, posinf=1.0, neginf=0.0)
        print("Image min:", np_sr_image.min(), "max:", np_sr_image.max())
        sr_img = (np.clip(np_sr_image, 0, 1) * 255).astype(np.uint8)
        sr_img = unpad_image(sr_img, pad_size*scale)
        sr_img = Image.fromarray(sr_img)

        return sr_img`

**Comment by noaft** [comment]
Thanks sir, my code worked with this code.
